### DISCUSSION TITLE
How to send event from Promise Actor to parent machine

### DISCUSSION THREAD

**Discussion by bflemi3** [statelyai/xstate#4684] [Q&A] [answered]
Hi All,

I'm new to xstate, and while the learning curve is steep, it's a really well thought-out / implemented library. I've been enjoying using it.

I'm working on a machine to handle uploading a file. During uploading, I'd like to show the user the progress. Here's what I have so far...

**machine.tsx**
```typescript
import upload from './actors/upload'

const machine = createMachine({
  id: 'Document Upload Machine',
  context: ({ input: { file } }) => ({
    created: Date.now(),
    file,
    name: getDocumentTitle(file.name),
  }),
  initial: 'init',
  states: {
    init: {},
    uploading: {
      entry: 'resetProgress',
      invoke: {
        src: 'upload',
        input: ({ context: { file, upload } }) => ({ file, urls: upload.urls }),
        onDone: {
          target: 'startProcessing',
          actions: assign({ upload: ({ context, event }) => ({ ...context.upload, parts: event.output }) }),
        },
        onError: {
          target: 'failed',
          actions: assign({ error: 'upload' }),
        },
      },
      on: {
        PROGRESS: {
          actions: {
            type: 'setProgress',
            params: ({ event }) => ({ value: event.value }) 
          }
        }
      }
    },
    ...
  }
}, {
  actions: {
    resetProgress: assign({ progress: 0 }),
    setProgress: assign({ progress: (_, { value }) => Math.round(value) }),
  },
  actors: {
    upload
  }
})
```
**./actors/upload.tsx**
```typescript
interface Input {
  file: File
  urls: string[]
}

const upload = fromPromise<IMultipartPart[], Input>(async ({ input: { file, urls }, self }) => {
  const partCount = urls.length
  const total = file.size
  const overallProgress = Array.from({ length: partCount }).map(() => 0)

  const _onUploadProgress = (part: number) => (event: ProgressEvent) => {
    overallProgress[part - 1] = event.loaded
    const loaded = overallProgress.reduce((loaded, part) => loaded + part, 0)

    console.log('Sending progress to parent', (loaded * 100) / total)

    sendParent({ type: 'PROGRESS', value: (loaded * 100) / total })
  }

  const parts = await Promise.all(
    urls.map(async (url, index) => {
      const partNumber = index + 1

      return uploadPart(
        url,
        { file, partCount, partNumber },
        { onUploadProgress: _onUploadProgress(partNumber) }
      )
    })
  )

  return parts
})
```

While I'm seeing `Sending progress to parent` in the console, a breakpoint in the `PROGRESS` event for the `uploading` state is not hit, showing that the event is not being picked up by the parent machine... and so the progress never gets updated. I've scoured the v5 documentation, the jsdocs site for the API, the xstate GitHub issues and pretty much exhausted google searching for an answer on how to properly send events from an Actor to Parent. 

To my understanding, a Promise Actor can send events to its parent via `sendParent`. What am I doing wrong or misunderstanding?

**Comment by nightire** [comment, marked as answer]
`sendParent` is not recommended in v5 any longer; I'd suggest reading the changelog to better understand the reason behind here: https://stately.ai/blog/2023-12-01-xstate-v5#reference-actors-anywhere.

There are generally two ways to send events from the promis actor back to its parent, either using a ref or using the actor system:

## By passing a ref

1. Invoking the promise actor with its parent's ref by `input`:

```ts
invoke: {
  src: 'upload',
  input: ({ self }) => ({ parent: self })
  ...
}
```

2. Send events

```ts
fromPromise(({ input: { parent } }) => {
  // ...
  parent.send({ type: 'PROGRESS' })
})
```

## By using the actor system

Read the docs here: https://stately.ai/docs/system

I recommend using the actor system because of its simplicity.

**Reply by bflemi3** [reply]
Much appreciated! I'll take a look at the actor system. Thank you for responding so quickly. 

The v5 documentation could be improved. I've spent all week reading through it, and it wasn't apparent to me that `sendParent` should no longer be used. The documentation does a good job of outlining components of the system, but IMO, lacks organization around HOW to do something (ie: How to communicate to a parent). Instead, the how-to details are buried in the System component of the docs, for instance.

For someone that's new to xstate, knowing the components of the library is great, but having documentation more organized around how to use the library (and those components) would have gotten me up and running much faster.

All that said, now that I'm understanding (mostly) how to use xstate, I'm really loving it.

**Reply by davidkpiano** [reply]
Thanks for the feedback. Where do you think that documentation on a `sendParent` alternative would be most appropriate?

**Reply by bflemi3** [reply]
Perhaps under Core Concepts > Actor Model > [Communication with events](https://stately.ai/docs/actor-model#communication-with-events). That section links to [Events and transitions](https://stately.ai/docs/transitions), so it could make sense to go there as well IMO.

**Reply by RemyMachado** [reply]
I have the exact same use case (uploading a file asynchronously and tracking progress).

I understand how passing a ref might work, but I have no idea how it would be done using the actor system.
Also, the ref approach doesn't feel ideal.

@davidkpiano would you mind explaining how to properly use the actor system for this very use-case please?

I have a very simple example using `fromCallback`, but I need it to switch to `fromPromise` because it's asynchronous:

```ts
   actions: {
      [ACTION.SET_UPLOAD_PROGRESS]: assign({
         uploadProgress: ({ event }) => {
            assertEvent(event, EVENT.FILES.SET_UPLOAD_PROGRESS);
            return event.progress;
         },
      }),
   },
   actors: {
      [ACTOR.UPLOAD_FILES_TO_S3]: fromCallback<EventObject, { files: File[] }>(
         ({ sendBack, input }) => {
            upload({
               files: input.files,
               onProgress: (progress) => {
                  sendBack({ type: EVENT.FILES.SET_UPLOAD_PROGRESS, progress });
               },
            });
         }
      ),
   },

**Reply by RemyMachado** [reply]
Also, when using the parent's ref approach, I'm not sure what the type should be... Is there a way to correctly type the input with `self` object?
